src/e2e/mod.rs

```diff
@@ -0,0 +1,156 @@
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("invalid bip39 phrase")]
+    InvalidPhrase,
+    #[error("invalid seed")]
+    InvalidSeed,
+    #[error("invalid cipher data")]
+    InvalidCipher,
+    #[error("k256: {0}")]
+    Secp256(#[from] secp256k1::Error),
+}
+
+/// Pair is a secure key (pair) that is used for encryption
+///
+pub struct Pair(KeyPair);
+
+impl FromStr for Pair {
+    type Err = Error;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let secp = Secp256k1::new();
+        let kp: KeyPair = match s.strip_prefix("0x") {
+            None => {
+                // no prefix, we assume this is a bip39 Mnemonic
+                let mnemonic = Mnemonic::from_phrase(s, Language::English)
+                    .map_err(|_| Error::InvalidPhrase)?;
+
+                let seed = bip39::Seed::new(&mnemonic, "");
+                // note: the secpk seed is only 32 bytes, so we take the first 32 bytes
+                // from the seed.
+                if seed.as_bytes().len() < 32 {
+                    return Err(Error::InvalidPhrase);
+                }
+
+                KeyPair::from_seckey_slice(&secp, &seed.as_bytes()[..32])?
+            }
+            Some(h) => {
+                let seed = hex::decode(h).map_err(|_| Error::InvalidSeed)?;
+                KeyPair::from_seckey_slice(&secp, &seed)?
+            }
+        };
+
+        Ok(Self(kp))
+    }
+}
+
+impl Pair {
+    /// get public key of the key pair. it's okay to share this public key
+    /// with other peers. actually it's the only way to use this for encryption
+    /// is by making your public key accessable for other peers
+    pub fn public(&self) -> [u8; PUBLIC_KEY_SIZE] {
+        self.0.public_key().serialize()
+    }
+
+    /// compute a "shared" key (used for encryption) knowing the public key of the other peer
+    /// Two peers A, and B can agree on the same shared key by sharing their own public key
+    /// hence
+    ///  shared1 = A.sk.shared(B.pk)
+    ///  shared2 = B.sk.shared(A.pk)
+    ///  assert_eq(shared1, shared2) <- is valid
+    pub(crate) fn shared<K: AsRef<[u8]>>(&self, pk: K) -> Result<[u8; SHARED_KEY_SIZE], Error> {
+        let pk = PublicKey::from_slice(pk.as_ref())?;
+
+        let point = secp256k1::ecdh::shared_secret_point(&pk, &self.0.secret_key());
+        use sha2::{Digest, Sha256};
+
+        let mut sh = Sha256::new();
+        sh.update(point);
+        Ok(sh.finalize().into())
+    }
+
+    /// Given two peers A, and B. Peer A can encrypt the data in a way that only B can decrypt it
+    /// by driving a shared key (check Pair::shared)
+    pub fn encrypt<K: AsRef<[u8]>, D: AsRef<[u8]>>(
+        &self,
+        pk: K,
+        data: D,
+    ) -> Result<Vec<u8>, Error> {
+        let shared = self.shared(&pk)?;
+        let mut nonce = [0u8; NONCE_KEY_SIZE];
+        OsRng.fill_bytes(&mut nonce);
+
+        let aes = Aes256Gcm::new_from_slice(&shared).unwrap();
```

Comment:
is this guaranteed not to fail?

Reply:
Yes, the only way this method can fail if the `key` (called shared in this case) has the wrong length! but the key is granted to have the correct length since it's an array with exact key length for Aes